modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileIndexerBoundsHelper.java

```diff
@@ -0,0 +1,325 @@
+     * bytes per envelope).
+     */
+    private static final int FALLBACK_TO_FILE_RECCOUNT_THRESHOLD = (1024 * 1024) / 32;
+
+    /**
+     * Returns a bounds reader strategy following the logic explained at the class' javadocs.
+     *
+     * @param reader the shapefile reader used to build the QuadTree
+     * @param shpIndex provided shapefile record count, and file offsets if necesasry
+     * @return a reader optimized to quickly store and serve the shapefile records bounds, may fall
+     *     back to random disk I/O on the {@code .shp} itself through {@link ShapefileReader}.
+     */
+    static BoundsReader createBoundsReader(final ShapefileReader reader, final IndexFile shpIndex) {
+        final int numRecs = shpIndex.getRecordCount();
+        final ShapeType shapeType = reader.getHeader().getShapeType();
+        final boolean pointBounds =
+                shapeType == POINT || shapeType == POINTM || shapeType == POINTZ;
+        final int fileRecCountThreshold =
+                (pointBounds ? 2 : 1) * FALLBACK_TO_FILE_RECCOUNT_THRESHOLD;
+
+        if (numRecs <= fileRecCountThreshold) {
+            return new HeapBoundsReader(numRecs, pointBounds);
+        }
+        try {
+            return new FileBoundsReader(numRecs, pointBounds);
+        } catch (IOException e) {
+            LOGGER.log(
+                    Level.INFO,
+                    "Unable to create a temporary file backed bounds reader, falling back to slower ShapefileReader randome access",
+                    e);
+            return new ShapefileReaderBoundsReader(reader, shpIndex);
+        }
+    }
+
+    /**
+     * Strategy to save the {@code .shp} records bounds while building the {@link QuadTree} and
+     * quickly access them when optimizing its layout.
+     *
+     * <p>While building the in memory {@link Quadtree}, the {@link ShapefileReader} is traversed
+     * sequentially. At that point, {@link #insert(int, Envelope)} should be called with each
+     * record's index and bounds, for this strategy object to save the bounds in its internal
+     * storage.
+     *
+     * <p>While optimizing the {@link QuadTree} layout, {@link #read(int, Envelope)} and {@link
+     * #expand(int, Envelope)} provided quick access to each record bounds.
+     *
+     * <p>Finally, {@link #close()} must be called to release any internal resources being used.
+     */
+    static interface BoundsReader {
+
+        /** Records the bounds of the given shapefile record to be read after. */
+        void insert(int recNumber, Envelope env);
+
+        /** Initializes the provided envelope to the bounds of the requested shapefile record. */
+        void read(int recNumber, Envelope env) throws IOException;
+
+        /** Expands the provided envelope to cover the bounds of the shapefile record number. */
+        void expand(int recNumber, Envelope env) throws IOException;
```

Comment:
Would recommend renaming to "expandEnvelope", this name seems like it would expand the reader somehow instead (at first I was looking for code that was growing the backing buffer/file, which did not make sense...)